File: runtime/src/langtype.c
#include "runtime_internal.h"

/* ── 签名 langtype（runtime篇-07，修订：无家族简写）──────────────────
 * type   = atom ("|" atom)*
 * atom   = [dims] ( any | kind )
 * dims   = "[]" | "[" elem ("," elem)* "]"   # 至多一个变元量词 ? 星 加
 * elem   = integer | "." | "?" | 星 | 加     # 轴量词（借鉴正则的 . 及量词）
 * any    = "any"           # 通配，匹配任意 kind
 * kind   = 精确 kind 串    # 见 known_kind
 *
 * 变参 "..." 是签名层 arity（吸收 0..N 个实参），不是 langtype 的一部分——
 * 由主槽 body 的 dynamic 字节承载，本文件的 valid/match 永不见 "..."。
 *
 * 铁律：不提供 int/uint/float/num 数值家族简写（位宽开放，纳入须逐个显式命名，
 * 如低精度浮点 float16/bfloat16/float8/e4m3/float8/e5m2；int4 等仍未纳入），
 * 也不提供 char 编码简写（编码须写明确，如 char/utf8、char/utf32）。
 * 多态靠显式 "|" 枚举（如 int8|int16|int32|int64）。
 */

static bool kind_eq(const char *s, size_t len, const char *k) {
    size_t kl = strlen(k);
    return len == kl && memcmp(s, k, len) == 0;
}

/* 精确 kind 集合（对齐 runtime kind 常量，不含 None）。 */
static bool known_kind(const char *s, size_t len) {
    return kind_eq(s, len, KVSPACE_KIND_BOOL) ||
           kind_eq(s, len, KVSPACE_KIND_INT8) || kind_eq(s, len, KVSPACE_KIND_INT16) ||
           kind_eq(s, len, KVSPACE_KIND_INT32) || kind_eq(s, len, KVSPACE_KIND_INT64) ||
           kind_eq(s, len, KVSPACE_KIND_UINT8) || kind_eq(s, len, KVSPACE_KIND_UINT16) ||
           kind_eq(s, len, KVSPACE_KIND_UINT32) || kind_eq(s, len, KVSPACE_KIND_UINT64) ||
           kind_eq(s, len, KVSPACE_KIND_FLOAT32) || kind_eq(s, len, KVSPACE_KIND_FLOAT64) ||
           kind_eq(s, len, KVSPACE_KIND_FLOAT16) || kind_eq(s, len, KVSPACE_KIND_BFLOAT16) ||
           kind_eq(s, len, KVSPACE_KIND_FLOAT8_E4M3) || kind_eq(s, len, KVSPACE_KIND_FLOAT8_E5M2) ||
           kind_eq(s, len, KVSPACE_KIND_CHAR) || kind_eq(s, len, KVSPACE_KIND_CHAR_UTF8) ||
           kind_eq(s, len, KVSPACE_KIND_CHAR_ASCII) ||
           kind_eq(s, len, KVSPACE_KIND_MAP) ||
           kind_eq(s, len, KVSPACE_KIND_RWIR) || kind_eq(s, len, KVSPACE_KIND_RWFUNC) ||
           kind_eq(s, len, KVSPACE_KIND_SCOPE) || kind_eq(s, len, KVSPACE_KIND_DEF_STRUCT) ||
           kind_eq(s, len, KVSPACE_KIND_TIME) ||
           kind_eq(s, len, KVSPACE_KIND_DURATION);
}

/* structref = "/" path：指向 /lib 下 struct 定义节点的完整路径（实例 kind / 字段类型）。
 * 仅语法承认（`/` + 合法路径段），存在性/字段一致性留给 runtime 判定。 */
static bool valid_structref(const char *s, size_t len) {
    if (len < 2 || s[0] != '/') return false;
    size_t seg = 0;
    for (size_t i = 1; i < len; i++) {
        if (s[i] == '/') {
            if (seg == 0) return false;
            seg = 0;
            continue;
        }
        char c = s[i];
        if (!((c >= 'a' && c <= 'z') || (c >= 'A' && c <= 'Z') ||
              (c >= '0' && c <= '9') || c == '_'))
            return false;
        seg++;
    }
    return seg > 0;
}

/* base = any | kind（kind 为精确合法 kind 串） */
static bool valid_base(const char *s, size_t len) {
    if (len == 0) return false;
    if (len == 3 && strncmp(s, "any", 3) == 0) return true;
    return known_kind(s, len);
}

/* elem = integer | "." | "?" | "*" | "+"（借鉴正则的轴量词）。 */
static bool elem_is_variable(const char *s, size_t len) {
    return len == 1 && (s[0] == '?' || s[0] == '*' || s[0] == '+');
}

static bool valid_elem(const char *s, size_t len) {
    if (len == 0) return false;
    if (len == 1 && (s[0] == '.' || s[0] == '?' || s[0] == '*' || s[0] == '+')) return true;
    for (size_t i = 0; i < len; i++)
        if (s[i] < '0' || s[i] > '9') return false;
    return true;
}

/* dims = ε（空 [] = 1 维任意，等价 [.]） | elem ("," elem)*；至多一个变元量词。 */
static bool valid_dims(const char *s, size_t len) {
    if (len == 0) return true;
    const char *p = s, *end = s + len;
    int nvar = 0;
    while (p < end) {
        const char *comma = memchr(p, ',', (size_t)(end - p));
        size_t seg = comma ? (size_t)(comma - p) : (size_t)(end - p);
        if (!valid_elem(p, seg)) return false;
        if (elem_is_variable(p, seg) && ++nvar > 1) return false;
        p += seg + (comma ? 1 : 0);
    }
    return true;
}

/* atom = [dims] base | structref */
static bool valid_atom(const char *s, size_t len) {
    if (len == 0) return false;
    if (s[0] == '/') return valid_structref(s, len);
    const char *p = s;
    if (*p == '[') {
        const char *end = memchr(p, ']', len);
        if (!end) return false;
        if (!valid_dims(p + 1, (size_t)(end - p - 1))) return false;
        p = end + 1;
        if (p >= s + len) return false;   /* 缺 base */
    }
    return valid_base(p, (size_t)(s + len - p));
}

/* 类型表达式语法校验（装载期）。变参 "..." 是签名层 arity、不入 langtype 串，此处永不见。 */
bool kvlangLangtypeValid(const char *expr) {
    if (!expr || !*expr) return false;
    const char *p = expr;
    const char *end = expr + strlen(expr);
    for (;;) {
        const char *pipe = memchr(p, '|', (size_t)(end - p));
        size_t len = pipe ? (size_t)(pipe - p) : (size_t)(end - p);
        if (!valid_atom(p, len)) return false;
        if (!pipe) break;
        p = pipe + 1;
        if (p >= end) return false;   /* 尾随 '|' → 空 atom */
    }
    return true;
}

/* any/kind 判定（kind 为运行时实际落盘 kind 串）。 */
static bool base_match(const char *s, size_t len, const char *kind) {
    if (len == 3 && strncmp(s, "any", 3) == 0) return true;
    return kind_eq(s, len, kind);
}

/* 定长元（整数或 "."）匹配单轴：. 任意长，整数须精确相等。 */
static bool fixed_elem_match(const char *s, size_t len, int32_t d) {
    if (len == 1 && s[0] == '.') return true;
    long v = 0;
    for (size_t j = 0; j < len; j++) v = v * 10 + (s[j] - '0');
    return v == d;
}

/* match_shape：轴量词序列 → (ndim,dims)。空串等价 "[.]"（恰一维）。
 * 至多一个变元量词，切分唯一无回溯：设定长元共 f 个，rem = ndim - f。 */
static bool match_shape(const char *s, size_t len, int32_t ndim, const int32_t *dims) {
    if (len == 0) return ndim == 1;
    /* 切分为 elem 段（最多 64 维）。 */
    const char *seg[64]; size_t seglen[64]; int m = 0;
    const char *p = s, *end = s + len;
    int vpos = -1;
    while (p < end && m < 64) {
        const char *comma = memchr(p, ',', (size_t)(end - p));
        size_t l = comma ? (size_t)(comma - p) : (size_t)(end - p);
        seg[m] = p; seglen[m] = l;
        if (elem_is_variable(p, l)) vpos = m;
        m++;
        p += l + (comma ? 1 : 0);
    }
    if (vpos < 0) {
        if (ndim != m) return false;
        for (int i = 0; i < m; i++)
            if (!fixed_elem_match(seg[i], seglen[i], dims[i])) return false;
        return true;
    }
    int f = m - 1;               /* 定长元个数 */
    int rem = ndim - f;          /* 变元量词吸收的轴数 */
    char q = seg[vpos][0];
    if (q == '?' && !(rem == 0 || rem == 1)) return false;
    if (q == '*' && rem < 0) return false;
    if (q == '+' && rem < 1) return false;
    for (int i = 0; i < vpos; i++)                    /* 前缀定长元 → dims[0..vpos] */
        if (!fixed_elem_match(seg[i], seglen[i], dims[i])) return false;
    int suf = m - 1 - vpos;                           /* 后缀定长元 → dims 末 suf 轴 */
    for (int k = 0; k < suf; k++)
        if (!fixed_elem_match(seg[vpos + 1 + k], seglen[vpos + 1 + k], dims[ndim - suf + k]))
            return false;
    return true;
}

/* match_atom：裸 kind 表单值（shape=[1]，即 ndim==0）；[dims] 前缀强制 shape。
 * 数组/字符串须显式 []kind（[]⟺[?] 一维）或 [n]kind。 */
static bool match_atom(const char *s, size_t len, const char *kind, int32_t ndim, const int32_t *dims) {
    if (len == 3 && strncmp(s, "any", 3) == 0) return true;   /* any：顶类型，任意 kind + 任意 shape */
    if (s[0] == '[') {
        const char *end = memchr(s, ']', len);
        if (!end) return false;
        if (!match_shape(s + 1, (size_t)(end - s - 1), ndim, dims)) return false;
        return base_match(end + 1, (size_t)(s + len - end - 1), kind);
    }
    return base_match(s, len, kind) && ndim == 0;
}

/* 单值（kind/ndim/dims）是否匹配类型表达式：任一 atom 命中即 true。
 * 变参 arity 不在此判（靠主槽 dynamic 字节 + 派发循环）。 */
bool kvlangLangtypeMatch(const char *expr, const char *kind, int32_t ndim, const int32_t *dims) {
    if (!expr || !kind) return false;
    const char *end = expr + strlen(expr);
    const char *p = expr;
    while (p < end) {
        const char *pipe = memchr(p, '|', (size_t)(end - p));
        size_t len = pipe ? (size_t)(pipe - p) : (size_t)(end - p);
        if (match_atom(p, len, kind, ndim, dims)) return true;
        if (!pipe) break;
        p = pipe + 1;
    }
    return false;
}

File: layout/src/langtype.rs
//! 签名 langtype（runtime篇-07，修订：无家族简写）：语法校验 + 值匹配。
//!
//! type   = atom ("|" atom)*
//! atom   = shape | mapexpr
//! shape  = [dims] ( any | kind )
//! mapexpr= key "·" type      # stringkeymap；value 递归 → 嵌套 map[string]map[…]
//! key    = "[]" char-enc     # 字符串键（= Go map[string]V）
//!        | "[" scalar ("," scalar)* "]"   # 标量元组键（物理以字符串格式落 key）
//! dims   = "[]" | "[" elem ("," elem)* "]"   # 至多一个变元量词 ?/*/+
//! elem   = integer | "." | "?" | "*" | "+"   # 轴量词（借鉴正则：. .? .* .+）
//! any    = "any"           # 通配，匹配任意 kind
//! kind   = 精确 kind 串    # 见 [`known_kind`]
//!
//! 铁律：不提供 int/uint/float/num 数值家族简写（位宽开放，纳入须逐个显式命名，
//! 如低精度浮点 float16/bfloat16/float8/e4m3/float8/e5m2；int4 等仍未纳入），
//! 也不提供 char 编码简写（编码须写明确，如 char/utf8、char/utf32）。多态靠显式 "|" 枚举。
//! 变参 "..." 是签名层 arity（吸收 0..N 个实参），不是 langtype 的一部分——由主槽 dynamic 字节承载。

/// 精确 kind 集合（对齐 runtime kind 常量，不含 None）。
fn known_kind(k: &str) -> bool {
    matches!(
        k,
        "bool"
            | "int8"
            | "int16"
            | "int32"
            | "int64"
            | "uint8"
            | "uint16"
            | "uint32"
            | "uint64"
            | "float32"
            | "float64"
            | "float16"
            | "bfloat16"
            | "float8/e4m3"
            | "float8/e5m2"
            | "char/utf32"
            | "char/utf8"
            | "char/ascii"
            | "stringkeymap"
            | "rwir"
            | "rwfunc"
            | "scope"
            | "struct"
            | "time"
            | "duration"
    )
}

/// structref = "/" path：指向 /lib 下类型定义节点的完整路径（实例 langtype、struct 字段类型）。
/// 仅做语法承认（`/` + 合法路径段），不解析 /lib 是否存在该类型 —— 存在性/字段一致性留给 runtime。
fn valid_structref(s: &str) -> bool {
    let rest = match s.strip_prefix('/') {
        Some(r) => r,
        None => return false,
    };
    // 段名字符集与**标识符**同一套（见 scanner::is_token_delim）——否则 `点` 这类 CJK
    // struct 名能作标识符、能作 struct 名，却单单不能出现在参数/返回的类型位置。
    !rest.is_empty()
        && rest.split('/').all(|seg| {
            !seg.is_empty()
                && seg
                    .bytes()
                    .all(|b| !super::scanner::is_token_delim(b) && b != b'"' && b != b'\'')
        })
}

fn valid_base(s: &str) -> bool {
    if s.is_empty() {
        return false;
    }
    s == "any" || known_kind(s)
}

fn elem_is_variable(s: &str) -> bool {
    s == "?" || s == "*" || s == "+"
}

/// elem = integer | "." | "?" | "*" | "+"（借鉴正则的轴量词）。
fn valid_elem(s: &str) -> bool {
    if s.is_empty() {
        return false;
    }
    matches!(s, "." | "?" | "*" | "+") || s.bytes().all(|b| b.is_ascii_digit())
}

/// dims = ε（空 [] = 1 维任意，等价 [.]） | elem ("," elem)*；至多一个变元量词 ?/*/+。
fn valid_dims(s: &str) -> bool {
    if s.is_empty() {
        return true;
    }
    let mut nvar = 0;
    for e in s.split(',') {
        if !valid_elem(e) {
            return false;
        }
        if elem_is_variable(e) {
            nvar += 1;
            if nvar > 1 {
                return false;
            }
        }
    }
    true
}

fn valid_shape(s: &str) -> bool {
    if s.is_empty() {
        return false;
    }
    if let Some(rest) = s.strip_prefix('[') {
        let end = match rest.find(']') {
            Some(e) => e,
            None => return false,
        };
        if !valid_dims(&rest[..end]) {
            return false;
        }
        let base = &rest[end + 1..];
        return !base.is_empty() && valid_base(base);
    }
    valid_base(s)
}

/// 标量 kind（可作元组键元素；不含 char/stringkeymap/index 等）。
fn valid_scalar(s: &str) -> bool {
    matches!(
        s,
        "bool"
            | "int8"
            | "int16"
            | "int32"
            | "int64"
            | "uint8"
            | "uint16"
            | "uint32"
            | "uint64"
            | "float32"
            | "float64"
    )
}

/// stringkeymap 的 key 恒 `[…]` 起头，两式：
///   `[]char/<enc>`      —— 字符串键（= Go `map[string]V`）
///   `[T1,T2,…]`         —— 标量元组键（元素为标量 kind，物理以字符串格式落 key）
fn valid_key(s: &str) -> bool {
    // memitemkey 三选一（见 [[map容器]]）：裸标量（`int64`）、字符串键 `[]char/<enc>`、
    // 标量元组键 `[scalar,…]`。裸标量必须认——`b:int64·int64 = {}` 是 spec 的标准写法。
    if valid_scalar(s) {
        return true;
    }
    let rest = match s.strip_prefix('[') {
        Some(r) => r,
        None => return false,
    };
    let end = match rest.find(']') {
        Some(e) => e,
        None => return false,
    };
    let inner = &rest[..end];
    let tail = &rest[end + 1..];
    if inner.is_empty() {
        return matches!(tail, "char/utf8" | "char/utf32" | "char/ascii");
    }
    tail.is_empty() && inner.split(',').all(valid_scalar)
}

/// atom = shape | mapexpr | structref；mapexpr = key "·" type；structref = "/" path。
/// 最前 `*` 是 ref 前缀（Ptr 存储位置），校验剥离后剩余部分（见 [[类型表达式文法]]）。
fn valid_atom(s: &str) -> bool {
    // `*`=ref ptr、`@`=ref @ext（见 [[ref存储位置]]）：两者都只是**源码**前缀，
    // layout 解析时剥离并落成 head.ref 字节，wire langtype 不含前缀。
    if let Some(rest) = s.strip_prefix('*').or_else(|| s.strip_prefix('@')) {
        return !rest.is_empty() && valid_atom(rest);
    }
    if s.starts_with('/') {
        return valid_structref(s);
    }
    if let Some(i) = s.find('·') {
        return valid_key(&s[..i]) && valid_langtype(&s[i + '·'.len_utf8()..]);
    }
    valid_shape(s)
}

/// 类型表达式语法校验（装载期）。变参 `...` 是签名层 arity、不入 langtype 串，此处永不见。
pub fn valid_langtype(expr: &str) -> bool {
    !expr.is_empty() && expr.split('|').all(valid_atom)
}

/// 是否只由**已知种类名**（或 `any`）构成——即「标量字面量可写入的类型」。
/// `*`/`@` 前缀先剥（源码传递方式，不是类型本体）；structref（`/lib/…`）与形状、mapexpr 皆否。
/// 用途见 `parser`（局部声明的写目标、struct 字段默认值）：裸名 `int`/`intg64` 经
/// [`expand_struct_refs`] 变成 `/lib/int` 后，正是靠这条落网——它**不是**种类名，标量写不进去
/// （见 [[文法与合法性]]）。
pub fn is_plain_kind(ty: &str) -> bool {
    !ty.is_empty()
        && ty.split('|').all(|a| {
            let a = a.trim_start_matches(['*', '@']);
            a == "any" || known_kind(a)
        })
}

/// 隐式 struct 名解析：把 langtype 中裸 struct 名（非 known kind / any 的标识符）展开为
/// `/lib/<name>`，使 kv 源可写 `x:Node` / `[int64]·Node`，runtime 恒收到完整 `/lib/…` 路径。
/// 已 `/` 开头或 known kind 原样返回。mapexpr 只对 value 递归展开（key 恒 `[…]` 非 struct）。
pub fn expand_struct_refs(s: &str) -> String {
    s.split('|').map(expand_atom).collect::<Vec<_>>().join("|")
}

fn expand_atom(s: &str) -> String {
    if let Some(rest) = s.strip_prefix('*') {
        return format!("*{}", expand_atom(rest));
    }
    if let Some(rest) = s.strip_prefix('@') {
        return format!("@{}", expand_atom(rest));
    }
    if s.starts_with('/') {
        return s.to_string();
    }
    // mapexpr `key·value`：键可以是 `[` 起头（字符串键/标量元组键）或**裸标量**，
    // 只有 value 侧需要展开 struct 名；整串当 structref 包成 `/lib/<key·value>` 是错的。
    if let Some(i) = s.find('·') {
        let key = &s[..i];
        if valid_key(key) {
            return format!("{key}·{}", expand_struct_refs(&s[i + '·'.len_utf8()..]));
        }
    }
    if s.starts_with('[') {
        return s.to_string();
    }
    if s.is_empty() || known_kind(s) || s == "any" {
        return s.to_string();
    }
    format!("/lib/{s}")
}

fn base_match(s: &str, kind: &str) -> bool {
    match s {
        "any" => true,
        _ => s == kind,
    }
}

/// 定长元（整数或 `.`）匹配单轴：`.` 任意长，整数须精确相等。
fn fixed_elem_match(e: &str, d: i32) -> bool {
    e == "." || e.parse::<i32>().ok() == Some(d)
}

/// 轴量词序列 → (ndim,dims)。空串等价 `[.]`（恰一维）。至多一个变元量词，切分唯一无回溯。
fn match_shape(s: &str, ndim: i32, dims: &[i32]) -> bool {
    if s.is_empty() {
        return ndim == 1;
    }
    let elems: Vec<&str> = s.split(',').collect();
    let m = elems.len();
    let vpos = elems.iter().position(|e| elem_is_variable(e));
    match vpos {
        None => {
            if ndim as usize != m {
                return false;
            }
            elems
                .iter()
                .enumerate()
                .all(|(i, e)| fixed_elem_match(e, dims[i]))
        }
        Some(vpos) => {
            let f = (m - 1) as i32; // 定长元个数
            let rem = ndim - f; // 变元量词吸收的轴数
            let ok_rem = match elems[vpos] {
                "?" => rem == 0 || rem == 1,
                "*" => rem >= 0,
                _ => rem >= 1, // "+"
            };
            if !ok_rem {
                return false;
            }
            // 前缀定长元 → dims[0..vpos]
            for (i, e) in elems.iter().take(vpos).enumerate() {
                if !fixed_elem_match(e, dims[i]) {
                    return false;
                }
            }
            // 后缀定长元 → dims 末 suf 轴
            let suf = m - 1 - vpos;
            for k in 0..suf {
                if !fixed_elem_match(elems[vpos + 1 + k], dims[ndim as usize - suf + k]) {
                    return false;
                }
            }
            true
        }
    }
}

/// ndim = -1 表示「已消费 dims，不再判 ndim」（递归哨兵）。
fn match_atom(s: &str, kind: &str, ndim: i32, dims: &[i32]) -> bool {
    // mapexpr（含 `·`）：容器值的 kind 恒为 stringkeymap，key/value 型不在此判。
    if s.contains('·') {
        return kind == "stringkeymap";
    }
    if let Some(rest) = s.strip_prefix('[') {
        let end = match rest.find(']') {
            Some(e) => e,
            None => return false,
        };
        if !match_shape(&rest[..end], ndim, dims) {
            return false;
        }
        return match_atom(&rest[end + 1..], kind, -1, &[]);
    }
    if ndim >= 0 && ndim != 0 {
        return false;
    }
    base_match(s, kind)
}

/// 单值（kind/ndim/dims）是否匹配类型表达式：任一 atom 命中即 true。
/// 变参 arity 不在此判（靠主槽 dynamic 字节 + 派发循环）。
pub fn match_langtype(expr: &str, kind: &str, ndim: i32, dims: &[i32]) -> bool {
    !expr.is_empty()
        && expr
            .split('|')
            .any(|atom| match_atom(atom, kind, ndim, dims))
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn valid() {
        for e in [
            "int64",
            "uint8",
            "float32",
            "bool",
            "any",
            "char/utf8",
            "char/utf32",
            "char/ascii",
            "stringkeymap",
            "[]float32",
            "[2]float32",
            "[2,3]float32",
            "[2,3,4]float64",
            "[.,768]float32",
            "[.,.]int8",
            "[.]float32",
            "[*]float32",
            "[?,3]float32",
            "[+,3]float32",
            "[2,*]float32",
            "[512,+]float32",
            "int64|float64",
            "[2,3]float32|float32",
            "[]float32|[]float64",
            "bool|char/utf8",
            "int64|stringkeymap",
            "[]char/utf8·int64",
            "[]char/utf32·[]char/utf8",
            "[]char/utf8·[]char/utf8·int64",
            "[int32,int32]·[]char/utf8",
            "[float32,float32]·[]char/utf8",
            "[float32,int8,uint32]·int64",
            "struct",
            "/lib/Point",
            "/lib/geom/Point",
            "/lib/Node",
            "/lib/Point|/lib/Node",
            // `*`/`@` 是**源码**前缀（layout 剥离落 head.ref），书于类型标注最前——合法（见 [[文法与合法性]]）。
            "*int64",
            "@int64",
            "@[256,256]uint8",
            "*[int32,int32]·[]char/utf8",
        ] {
            assert!(valid_langtype(e), "{e} should be valid");
        }
        for e in [
            "·int64",
            "[]char/utf8·",
            "char/utf8·int64",
            "int·int64",
            "[]char/utf8·nope",
            "[]int32·int64",
            "[.]·int64",
            "[int32,]·int64",
            "[foo,int32]·int64",
            "any...",
            "int64|float64...",
            "[]float32...",
            "[*,+]int64",
            "[?,*]int64",
            "/",
            "/lib/",
            "/lib//Point",
            "lib/Point",
            "/lib/Po int",
        ] {
            assert!(!valid_langtype(e), "{e} should be invalid");
        }
        assert!(match_langtype("/lib/Point", "/lib/Point", 0, &[]));
        assert!(!match_langtype("/lib/Point", "/lib/Node", 0, &[]));
        assert!(match_langtype("[]char/utf8·int64", "stringkeymap", 1, &[3]));
    }

    #[test]
    fn quantifiers() {
        // "." 恰 1 轴任意长
        assert!(match_langtype("[.]int64", "int64", 1, &[5]));
        assert!(!match_langtype("[.]int64", "int64", 2, &[2, 3]));
        // "*" 0+ 轴
        assert!(match_langtype("[*]int64", "int64", 0, &[]));
        assert!(match_langtype("[*]int64", "int64", 3, &[2, 3, 4]));
        // "+" 1+ 轴
        assert!(!match_langtype("[+]int64", "int64", 0, &[]));
        assert!(match_langtype("[+]int64", "int64", 2, &[2, 3]));
        // "?" 0 或 1 轴
        assert!(match_langtype("[?,3]int64", "int64", 1, &[3]));
        assert!(match_langtype("[?,3]int64", "int64", 2, &[7, 3]));
        assert!(!match_langtype("[?,3]int64", "int64", 3, &[7, 8, 3]));
        // 前缀定长 + 尾量词
        assert!(match_langtype("[512,*]int64", "int64", 1, &[512]));
        assert!(match_langtype("[512,*]int64", "int64", 3, &[512, 8, 8]));
        assert!(!match_langtype("[512,*]int64", "int64", 2, &[7, 8]));
    }

    #[test]
    fn invalid() {
        for e in [
            "",
            "int|",
            "|int",
            "int||float64",
            "|",
            "[]",
            "[2]",
            "[?]",
            "[2",
            "2]",
            "[2,]float32",
            "[,2]float32",
            "[2 3]float32",
            "int64*",
            "float64|",
            "int ",
            "float32,float64",
            "int",
            "uint",
            "float",
            "num",
            "char",
            "int4",
            "fp8",
            "fp16",
            "string",
            "charbyte",
        ] {
            assert!(!valid_langtype(e), "{e} should be invalid");
        }
    }

    #[test]
    fn matching() {
        let cases = [
            ("int64", "int64", 0, &[][..], true),
            ("int64", "float64", 0, &[], false),
            ("any", "stringkeymap", 0, &[], true),
            ("any", "int4", 0, &[], true),
            ("char/utf8", "char/utf8", 0, &[], true),
            ("char/utf8", "char/utf32", 0, &[], false),
            ("int64|float64", "float64", 0, &[], true),
            ("int64|float64", "bool", 0, &[], false),
            ("[]float32", "float32", 1, &[5], true),
            ("[]float32", "float32", 2, &[2, 3], false),
            ("[]float32", "float32", 0, &[], false),
            ("[2,3]float32", "float32", 2, &[2, 3], true),
            ("[2,3]float32", "float32", 2, &[2, 4], false),
            ("[.,768]float32", "float32", 2, &[100, 768], true),
            ("[.,768]float32", "float32", 2, &[100, 512], false),
            ("[2,3]float32|float32", "float32", 0, &[], true),
            ("[2,3]float32|float32", "float64", 0, &[], false),
            ("[]float32|[]float64", "float64", 1, &[10], true),
            ("bool|char/utf8", "char/utf8", 0, &[], true),
            ("int64|stringkeymap", "int64", 0, &[], true),
        ];
        for (expr, kind, ndim, dims, want) in cases {
            let got = match_langtype(expr, kind, ndim, dims);
            assert_eq!(
                got, want,
                "match_langtype({expr}, {kind}, ndim={ndim}, dims={dims:?})"
            );
        }
    }
}
